### ISSUE TITLE
In private network, cpp-ethereum's cpu miner stops once a new node joining this network

### ISSUE THREAD

**Issue by azazhu** [ethereum/aleth#5091]
**OS:**
 two Ubuntu 16.04.4 LTS

**Issue:**
In private network, if there is only one node in network, everything works fine: node A can mine blocks and results can be accepted by itself. But once new node B joins this private network, node A's cpu miner stops. Is it a issue?

**Repro:**
1. launch node A, use geth as console and start mining. 2. launch node B, use geth as console and join node A's private network. 3. node A's miner stops

**Details:**
1. launch node A and start mining:
cmdline of cpp-ethereum:
```
./aleth --config /data/local/abc/opensource/orig_eth_cppclient/genesis.json -m on -v 5 --network-id 45676 --ipc --listen 30307 --port 30307 --db-path /data/local/abc/opensource/orig_eth_cppclient/data --private "eth_45676"
```
cmdline of go-ethereum:
```
./geth attach /data/local/abc/opensource/orig_eth_cppclient/data/geth.ipc
		>>miner.setEtherbase(eth.accounts[0])
		>>miner.start(1)
```
output of cpp-ethereum:
```
...
2018-06-22 17:33:38 eth trace Generating seal on#c1d77054…#36
2018-06-22 17:33:38 ethsync host Sending a new block (current is #372151a1…, was #7a003d2c…)
2018-06-22 17:33:38 miner2 block Sealing block!
2018-06-22 17:33:38 miner2 bq Queuing block #314aeb70… for import...
2018-06-22 17:33:38 miner2 blockhdr Expected trie root: 56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
2018-06-22 17:33:38 miner2 blockhdr Expected uncle hash: 1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347
2018-06-22 17:33:38 miner2 bq Block #314aeb70… is 36 parent is #372151a1…
2018-06-22 17:33:38 miner2 bq OK - ready for chain insertion.
2018-06-22 17:33:38 verifier5 blockhdr Expected trie root: 56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
2018-06-22 17:33:38 verifier5 blockhdr Expected uncle hash: 1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347
2018-06-22 17:33:38 eth blockhdr Expected trie root: 56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421
2018-06-22 17:33:38 eth blockhdr Expected uncle hash: 1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347
2018-06-22 17:33:38 eth chain Attempting import of #314aeb70… ...
2018-06-22 17:33:38 eth overlaydb Closing state DB
2018-06-22 17:33:38 eth block Committing to disk: stateRoot #e3103e1e… = #e3103e1e… = f8518080808080a00f571588cf7cc4b8566dc9464b916f01aed62ff1076524db905424bae3bec9aa808080a0a7c6beb192fcc3e67b492b7256805be7f5fd4bb2dd5a34deef92269aa6f22e8580808080808080
2018-06-22 17:33:38 eth block Committed: stateRoot #e3103e1e… = #e3103e1e… = f8518080808080a00f571588cf7cc4b8566dc9464b916f01aed62ff1076524db905424bae3bec9aa808080a0a7c6beb192fcc3e67b492b7256805be7f5fd4bb2dd5a34deef92269aa6f22e8580808080808080
2018-06-22 17:33:38 eth block finalising enactment. current -> previous, hash is #314aeb70…
2018-06-22 17:33:38 eth chain    Imported and best 4887747 (#36). Has 0 siblings. Route: [#372151a1…, #314aeb70…]
2018-06-22 17:33:38 eth client 1 blocks imported in 0 ms (1795.33 blocks/s) in #36
2018-06-22 17:33:38 eth client Live block: #314aeb70…
2018-06-22 17:33:38 eth overlaydb Closing state DB
2018-06-22 17:33:38 eth overlaydb Closing state DB
2018-06-22 17:33:38 eth client Post state changed.
2018-06-22 17:33:38 eth watch noteChanged: {chain}
2018-06-22 17:33:38 eth trace No transactions to process. 0 pending, 0 queued, 0 future, 0 unverified
2018-06-22 17:33:38 eth client Rejigging seal engine...
2018-06-22 17:33:38 eth block Checking #314aeb70…, parent = #372151a1…
2018-06-22 17:33:38 eth block Post-reward stateRoot: #9b8b87b2…
2018-06-22 17:33:38 eth block --- 9b8b87b208cf548642dc8996be295daff2796494be5f4dec230a8e0318af8cc6
     0000000000000000000000000000000000000000: 0 #:3000000000000000000 [SIMPLE]
     f9086b540e3c8e4b4aed10293f494b958691d358: 0 #:108000000000000000000 [SIMPLE]

2018-06-22 17:33:38 eth trace Generating seal on#bfd99b9f…#37
...
```

2.  launch node B and then join node A's network:
cpp-ethereum cmdline: 
`./aleth --config /home/data.abc_gpu/opensource/orig_eth_cppclient/genesis.json -m on -v 5 --network-id 45676 --ipc --listen 30307 --port 30307 --db-path /home/data.abc_gpu/opensource/orig_eth_cppclient/data --private "eth_45676"
`
go-ethereum cmdline: 
```
./geth attach /home/data.abc_gpu/opensource/orig_eth_cppclient/data/geth.ipc
admin.addPeer("enode://4c9e8fdcbd2a06c66d8ff080edb9abb366828370e0769749a642f6cc6e06c858bdb0f600f5a30d0dab2a40a141eb4eafa9eedc04e729e7e72e7f1a8c906d54d9@10.19.225.196:30307")
```		

3.  node A's miner stops
output of node A:
```
...
B6CCD41AD312451B948A7413F0A142FD40D49347, 0xF9086B540E3C8E4B4AED10293F494B958691D358, 0xEEAAB16BCBF15A0EF6B9635ADF5ABF520A4A297638678627CD2C629CE37C834A, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 0x22BB3, 0xAD, 0xD8303283, 0x0, 0x5B2CC34B, 0xD68094312E342E302B2B32323664726C696E7578676E75, 0xE78FAC5CC7192FC233EA46227E8CAEFF752E1D8ADD5BDA659C0656283E302AA8, 0x9625D1A8ED7A7180 ] ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 19 [ 0xAA540FDABC421C79D2187A05C96F16FABA9625702614F065A7B5DC1A8CB83629, 0x1, 0x0, 0x0 ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo GetBlockHeaders (block (hash): #aa540fda…, maxHeaders: 1, skip: 0, reverse: false)
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo <- [ [ 0xB7038ECCDEEBA1BB6080423899B57777F2E241CF8C9CB62FD44482D05C194BB1, 0x1DCC4DE8DEC75D7AAB85B567B6CCD41AD312451B948A7413F0A142FD40D49347, 0xF9086B540E3C8E4B4AED10293F494B958691D358, 0x1FB74EFF874139FCDBE6DBD0C4CC778D6FA9112B7C6EC56CB28F71DBEBD87C97, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 0x22BF8, 0xAE, 0xD7FA2678, 0x0, 0x5B2CC34C, 0xD68094312E342E302B2B32323664726C696E7578676E75, 0x5F683E0592F19755A6D58AD009C5E77CBB999DFF63B4F0177BBFCC6856D9C215, 0x56F1FED40AFA2AEF ] ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 19 [ 0x58, 0x1, 0x0, 0x0 ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo GetBlockHeaders (88 max: 1 skip: 0)
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo <- [ [ 0xDE2BF9DF1EC5BAF25895F361747D5FF0DBC647984824529FE76FA8FC2A3AA882, 0x1DCC4DE8DEC75D7AAB85B567B6CCD41AD312451B948A7413F0A142FD40D49347, 0xF9086B540E3C8E4B4AED10293F494B958691D358, 0xA02E702D69CA3FEBF35CC48E42CC70626917796283B80805109024BAB83722FF, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 0x21506, 0x58, 0xEAE8B0E1, 0x0, 0x5B2CC2F6, 0xD68094312E342E302B2B32323664726C696E7578676E75, 0x790FA5168B6CC8B18F6734143061C31871BD7A32454B519D4EAF6D2A71516C04, 0x7B3B5A7BA8D54823 ] ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 19 [ 0x59, 0x55, 0x0, 0x0 ]
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo GetBlockHeaders (89 max: 85 skip: 0)
2018-06-22 17:35:50 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo <- [ [ 0x21906CD2FD33EED93795E3375BF98CF6C4501674FB38AC52F468BC75C01ECE15, 0x1DCC4DE8DEC75D7AAB85B567B6CCD41AD312451B948A7413F0A142FD40D49347, 0xF9086B540E3C8E4B4AED10293F494B958691D358, 0xBE5CAAC8EA5C72F18D7AE5D505CED0CFE18A12539D4DCFA4B8DA6E5D1D33D35D, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, 0x56E81F171BCC55A6FF8345E692C0F86E5B48E01B996CADC001622FB5E363B421, "\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x
...
00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00", 0x22BB3, 0xAD, 0xD8303283, 0x0, 0x5B2CC34B, 0xD68094312E342E302B2B32323664726C696E7578676E75, 0xE78FAC5CC7192FC233EA46227E8CAEFF752E1D8ADD5BDA659C0656283E302AA8, 0x9625D1A8ED7A7180 ] ]
2018-06-22 17:35:52 eth client Since 2018-06-22 09:35:35.237Z (17): 15ticks
2018-06-22 17:35:57 p2p net <- [ ]
2018-06-22 17:35:57 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 3 [ ]
2018-06-22 17:35:57 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo Latency: 1 ms
2018-06-22 17:36:07 eth client Since 2018-06-22 09:35:52.657Z (15): 15ticks
2018-06-22 17:36:14 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 2 [ ]
2018-06-22 17:36:14 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo Ping ##874aa6af…
2018-06-22 17:36:14 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo <- [ ]
2018-06-22 17:36:22 eth client Since 2018-06-22 09:36:07.662Z (15): 15ticks
2018-06-22 17:36:27 p2p net <- [ ]
2018-06-22 17:36:27 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo -> 3 [ ]
2018-06-22 17:36:27 p2p net 874aa6af…|aleth/1.4.0.dev1-34+commit.226d834a/linux/gnu5.4.0/relwithdebinfo Latency: 1 ms
2018-06-22 17:36:37 eth client Since 2018-06-22 09:36:22.667Z (15): 15ticks
2018-06-22 17:36:41 aleth rpc 
...
```

4. check their blocknumber: node A's blocknumber is 175 and node b's blocknumber is 91

5. genesis.json
```
{  
	"sealEngine": "Ethash",  
	"params": {  
		"accountStartNonce": "0x00",  
		"maximumExtraDataSize": "0x20",  
		"homesteadForkBlock": "0x00",  
		"EIP150ForkBlock": "0x00",  
		"EIP158ForkBlock": "0x00",  
		"byzantiumForkBlock": "0x00",  
		"constantinopleForkBlock": "0x00",  
		"minGasLimit": "0x1388",  
		"maxGasLimit": "7fffffffffffffff",  
		"tieBreakingGas": false,  
		"gasLimitBoundDivisor": "0x0400",  
		"minimumDifficulty": "0x020000",  
		"difficultyBoundDivisor": "0x0800",  
		"durationLimit": "0x0d",  
		"blockReward": "0x4563918244F40000",  
		"networkID" : "0xB26C",  
		"chainID": "0x01",  
		"allowFutureBlocks" : true, 
		"daoHardforkBlock": "0x00"
	},  
	"genesis": {  
		"nonce": "0x0000000000000042",  
		"difficulty": "0x400000000",  
		"mixHash": "0x0000000000000000000000000000000000000000000000000000000000000000",  
		"author": "0x0000000000000000000000000000000000000000",  
		"timestamp": "0x00",  
		"parentHash": "0x0000000000000000000000000000000000000000000000000000000000000000",  
		"extraData": "0x11bbe8db4e347b4e8c937c1c8370e4b5ed33adb3db69cbdb7a38e1e50b1b82fa",  
		"gasLimit": "0x1388"  
	},  
	"accounts": {  
		"0000000000000000000000000000000000000001": { "precompiled": { "name": "ecrecover", "linear": { "base": 3000, "word": 0 } } },  
		"0000000000000000000000000000000000000002": { "precompiled": { "name": "sha256", "linear": { "base": 60, "word": 12 } } },  
		"0000000000000000000000000000000000000003": { "precompiled": { "name": "ripemd160", "linear": { "base": 600, "word": 120 } } },  
		"0000000000000000000000000000000000000004": { "precompiled": { "name": "identity", "linear": { "base": 15, "word": 3 } } },  
		"0000000000000000000000000000000000000005": { "precompiled": { "name": "modexp" } },  
		"0000000000000000000000000000000000000006": { "precompiled": { "name": "alt_bn128_G1_add", "linear": { "base": 500, "word": 0 } } },  
		"0000000000000000000000000000000000000007": { "precompiled": { "name": "alt_bn128_G1_mul", "linear": { "base": 40000, "word": 0 } } },  
		"0000000000000000000000000000000000000008": { "precompiled": { "name": "alt_bn128_pairing_product" } }  
	}  
}  

```

**Comment by gumb0** [comment]
Yeah, looks like for me sync is getting stuck soon after beginning on both nodes.

**Comment by ifdefelse** [comment]
@gumb0, do you mean 1) once sync starts, the miner stops but 2) you also found the sync get blocked on both nodes for unknown reason so 3) the miner has no chance to re-start as the sync won't finish. 
But in this situation, node A is the first node joining this network and mines blocks and node B just sync without mining. Why node A should wait for node B finish sync ? i think node A can continue mining, right?

**Comment by azazhu** [comment]
@gumb0, do you mean 1) once sync starts, the miner should stop but 2) you also found the sync get blocked on both nodes for unknown reason so 3) the miner has no chance to re-start as the sync won't finish.
But in this situation, node A is the first node joining this network and mines blocks and node B just sync without mining. Why node A should wait for node B finishing sync ? i think node A can continue mining when node B is syncing, right?

**Comment by gumb0** [comment]
Oh somehow I thought that both nodes are mining in your scenario, sorry.
I'll try with only node A mining, that should be easier to debug

**Comment by gumb0** [comment]
https://github.com/ethereum/cpp-ethereum/pull/5095 This should fix the stopping of the mining on the node A.

But the sync of the node B doesn't finish correctly for me (gets stuck), I will investigate it further.

**Comment by gumb0** [comment]
https://github.com/ethereum/cpp-ethereum/pull/5096 This should improve syncing on the node B

**Comment by azazhu** [comment]
Thx @gumb0 . I don't know the process. Once your fix code be reviewed, they would be submitted to master branch. I don't need to do anything(like verification), right?

**Comment by gumb0** [comment]
You don't have to, but I would appreciate if you check how it works for you.

I can ping you when we merge it, or you could try it from PR branches.

**Comment by azazhu** [comment]
@gumb0 I pulled your fixes and tried. They can fix pervious issues. But I hit another issue.
**Problem:**
Once Node A stops mining, Node B stops syncing. I have to re-launch Node B to make it sync again.
**Repro:**
1. launch Node A and Node B. Node B addPeer Node A.
2. Node A starts mining. Node B syncs Node A's blocks(but Node B is always behind Node A)
3. Node A stops mining. Node B stops sync(for sample, Node A is at 280 block and Node B gets stuck at 220 block)
4. Close and re-launch Node B. Node B can sync to highest block(both Node A & B reach 280 block)

**Comment by ifdefelse** [comment]
adding more info for #3: though Node B stops syncing, it know A's highest block.
type eth.syncing at node B. It output
>>eth.syncing
{
  currentBlock: 220,
  highestBlock: 280
  startingBlock: 51
}

**Comment by gumb0** [comment]
Thank you, I see this happening, too.

**Comment by gumb0** [comment]
https://github.com/ethereum/cpp-ethereum/pull/5099 This should further improve sync on Node B
(it shouldn't stay behind Node A as it did before, for me it syncs thousands of blocks pretty quickly)

**Comment by gumb0** [comment]
@azazhu Could you try how latest develop works for you?

**Comment by azazhu** [comment]
@gumb0 I tried but met a new issue.
**Problem:**
the miner stops after a few blocks.
**Repro:**
1.launch Node A and Node B. Node B addPeer Node A.
2.Node A starts mining. Node B syncs Node A's blocks. 
3.But Node A stops mining after a few blocks, and outputs:
``

**Comment by gumb0** [comment]
I see now some weird behavior when the sync starts on the mining node sometimes even when genesis hash of the peer is not equal to ours, i.e. comparing the genesis hashes doesn't work right (may be data races around hash variable). Node B is not needed to reproduce this.

**Comment by gumb0** [comment]
Now I don't think the reason is data races, but some incorrect logic in `BlockChainSync` which sometimes started the sync even when genesis hash is different

@azazhu Please check how this fix works for you https://github.com/ethereum/cpp-ethereum/pull/5109

**Comment by azazhu** [comment]
@gumb0 , I met a compiler error when build tot cpp-ethereum in ubuntu16.04. 
"Could not find toolchain file:
  /scratch/local/xxx/opensource/cpp-ethereum/evmc/cmake/cable/toolchains/cxx11-pic.cmake"
Do you have any idea what's wrong?

**Comment by gumb0** [comment]
@azazhu try `git submodule update --init --recursive`

**Comment by gumb0** [comment]
Hi @azazhu was your build issue resolved?

**Comment by ifdefelse** [comment]
Using the latest code from github, there are still issues:
**node A**: the miner in node A can mine and update client A's block number, while node B can receive node A's update
**node B**: node B's miner can not start, and note B's block number never get updated.

**Comment by davilizh** [comment]
Using the latest cpp-ethereum code from github, still meet similar issue:
**Node A**: miner can start successfully, blockNum get updated as mining, and node B can receive node A's update, bolck B's node Number never get changed.
**Node B**: miner cannot start and blockNum never get changed.

**Comment by gumb0** [comment]
Feel free to open another issue if you still experience problems
